pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/parser/GroovyShellDecoratorImpl.java

```diff
@@ -33,15 +46,34 @@ public void configureCompiler(@CheckForNull final CpsFlowExecution execution, Co
         cc.addCompilationCustomizers(new CompilationCustomizer(CompilePhase.SEMANTIC_ANALYSIS) {
             @Override
             public void call(SourceUnit source, GeneratorContext context, ClassNode classNode) throws CompilationFailedException {
-                // TODO: workflow-cps-plugin CpsFlowExecution.parseScript() should be passing in CodeSource
-                // to help us determine that that is a user-written script
-                // Commenting out for findbugs
-                // CodeSource cs = classNode.getCompileUnit().getCodeSource();
-                // if (<< cs comes from Jenkinsfile>>) {
-
-                // but until that happens,
-                // Using getNameWithoutPackage to make sure we don't end up executing without parsing when an inadvertent package name is put in the Jenkinsfile.
-                if (classNode.getNameWithoutPackage().equals(Converter.PIPELINE_SCRIPT_NAME)) {
+                boolean doModelParsing = false;
+
+                // First, we'll parse if this is WorkflowScript, i.e., a Jenkinsfile, or it's an autogenerated script
+                // with no package, i.e., Script1
+                if (classNode.getNameWithoutPackage().equals(Converter.PIPELINE_SCRIPT_NAME) ||
+                        (classNode.getPackageName() == null && classNode.getNameWithoutPackage().matches("Script\\d+"))) {
+                    doModelParsing = true;
+                } else if (execution != null && classNode.getPackageName() == null) {
+                    // Second, if we've got an execution and there's no package name, we'll parse if this is a global
+                    // variable, which we can determine by looking to see if its name is in the list of global variables.
+                    // Note that the combination of no package name plus global variable name should keep us from
+                    // parsing src/org/whatever/Foo.groovy in shared libraries *or* global variables defined in plugins.
+                    try {
+                        FlowExecutionOwner owner = execution.getOwner();
+                        if (owner != null && owner.getExecutable() instanceof WorkflowRun) {
+                            WorkflowRun run = (WorkflowRun)owner.getExecutable();
```

Comment:
`Run` suffices here

Reply:
Gotcha. At some point I was doing something that actually cared about `WorkflowRun`.